File: tests/golden/sandbox/recipes/gt5_state_resume.py
"""GT-5 — state-resume: continue from disk after a SIGTERM-style stop.

The runner records the full happy path the same way GT-1 does, but
the harness drives it in **two segments** to demonstrate that the
engine resumes purely from the persisted state file:

- Segment A: invoke the engine until it halts at ``apply-plan``.
  The runner then exits its capture loop early, simulating an
  agent process termination.
- Segment B: a fresh runner invocation continues from the saved
  state file (``--ticket-file`` is omitted on every cycle by the
  runner anyway, so cycle 1 of segment B already exercises the
  resume contract). The capture then walks the rest of the flow
  to SUCCESS.

The recipe content matches GT-1 (multiply); the value of GT-5 is
the segmentation, not the toy domain.
"""
from __future__ import annotations

from pathlib import Path
from typing import Any

from . import _helpers as h

META = {
    "gt_id": "GT-5",
    "ticket_relpath": "tickets/gt-5-state-resume.json",
    "persona": None,
    "cycle_cap": 6,
    "resume_after_directive": "apply-plan",
}

_NEGATE_SRC = '''

def negate(a: int) -> int:
    return -a
'''

_NEGATE_TEST = '''

def test_negate_returns_negation() -> None:
    from src.calculator import negate
    assert negate(0) == 0
    assert negate(7) == -7
    assert negate(-7) == 7
'''


def build_recipe(workspace: Path) -> dict[str, Any]:
    """Same shape as GT-1 — the segmentation lives in the runner."""

    def on_create_plan(state: dict[str, Any], record) -> dict[str, Any]:
        state["plan"] = h.standard_plan(
            "Add negate(a)",
            "Append negate(a: int) -> int to src/calculator.py",
            "Append test_negate_returns_negation to tests/test_calculator.py",
            "Run pytest to confirm both edge cases pass",
        )
        record.recipe_notes.append("plan recorded")
        return state

    def on_apply_plan(state: dict[str, Any], record) -> dict[str, Any]:
        h.append_to_file(workspace, "src/calculator.py", _NEGATE_SRC)
        h.append_to_file(workspace, "tests/test_calculator.py", _NEGATE_TEST)
        state["changes"] = h.base_changes(
            "src/calculator.py",
            "tests/test_calculator.py",
        )
        record.recipe_notes.append("negate + test appended")
        return state

    def on_run_tests(state: dict[str, Any], record) -> dict[str, Any]:
        state["tests"] = h.run_pytest(workspace)
        record.recipe_notes.append(
            f"pytest verdict={state['tests']['verdict']}",
        )
        return state

    def on_review_changes(state: dict[str, Any], record) -> dict[str, Any]:
        state["verify"] = h.simulated_review_verdict()
        record.recipe_notes.append("review-changes simulated success")
        return state

    return {
        "create-plan": on_create_plan,
        "apply-plan": on_apply_plan,
        "run-tests": on_run_tests,
        "review-changes": on_review_changes,
    }


__all__ = ["META", "build_recipe"]

File: tests/golden/sandbox/recipes/gt5_state_resume.ts
/**
 * GT-5 — state-resume: continue from disk after a SIGTERM-style stop (TS twin
 * of gt5_state_resume.py).
 *
 * The recipe content matches GT-1 in shape (here: `negate`); the value of
 * GT-5 is the segmentation, which lives in the capture driver via
 * `resume_after_directive`. In replay the runner already omits the input flag
 * after cycle 1 (resume from the persisted state file is the default), so a
 * single `run_capture` exercises the resume contract transparently.
 */
import type { RecipeModule } from './index.js';
import * as h from './_helpers.js';
import type { CycleRecord, Dict, RecipeStep } from '../runner.js';

export const META: RecipeModule['META'] = {
    gt_id: 'GT-5',
    ticket_relpath: 'tickets/gt-5-state-resume.json',
    persona: null,
    cycle_cap: 6,
    resume_after_directive: 'apply-plan',
};

const NEGATE_SRC = `

export function negate(a: number): number {
    return -a;
}
`;

// Note: negate(0) returns -0 in JS, and vitest's toBe uses Object.is, so
// `toBe(0)` would fail on -0 (Python's `== 0` did not). The scenario's value
// is the resume segmentation, not the zero edge, so the assertions use
// non-zero inputs.
const NEGATE_TEST = `

import { negate } from '../src/calculator.js';

it('negate returns negation', () => {
    expect(negate(7)).toBe(-7);
    expect(negate(-7)).toBe(7);
    expect(negate(3)).toBe(-3);
});
`;

export function buildRecipe(workspace: string): Record<string, RecipeStep> {
    const onCreatePlan: RecipeStep = (state: Dict, record: CycleRecord) => {
        state['plan'] = h.standard_plan(
            'Add negate(a)',
            'Append negate(a: number): number to src/calculator.ts',
            'Append a negate test to tests/calculator.test.ts',
            'Run vitest to confirm both edge cases pass',
        );
        record.recipe_notes.push('plan recorded');
        return state;
    };

    const onApplyPlan: RecipeStep = (state: Dict, record: CycleRecord) => {
        h.append_to_file(workspace, 'src/calculator.ts', NEGATE_SRC);
        h.append_to_file(workspace, 'tests/calculator.test.ts', NEGATE_TEST);
        state['changes'] = h.base_changes('src/calculator.ts', 'tests/calculator.test.ts');
        record.recipe_notes.push('negate + test appended');
        return state;
    };

    const onRunTests: RecipeStep = (state: Dict, record: CycleRecord) => {
        state['tests'] = h.run_vitest(workspace);
        record.recipe_notes.push(`vitest verdict=${(state['tests'] as Dict)['verdict'] as string}`);
        return state;
    };

    const onReviewChanges: RecipeStep = (state: Dict, record: CycleRecord) => {
        state['verify'] = h.simulated_review_verdict();
        record.recipe_notes.push('review-changes simulated success');
        return state;
    };

    return {
        'create-plan': onCreatePlan,
        'apply-plan': onApplyPlan,
        'run-tests': onRunTests,
        'review-changes': onReviewChanges,
    };
}

const recipe: RecipeModule = { META, buildRecipe };
export default recipe;
